editoast/src/views/timetable/stdcm/request.rs

```diff
@@ -0,0 +1,288 @@
+    /// Deprecated, first step data should be used instead
+    pub maximum_run_time: Option<u64>,
+    /// Train categories for speed limits
+    // TODO: rename the field and its description
+    pub speed_limit_tags: Option<String>,
+    /// Margin before the train passage in seconds
+    ///
+    /// Enforces that the path used by the train should be free and
+    /// available at least that many milliseconds before its passage.
+    #[serde(default)]
+    pub time_gap_before: u64,
+    /// Margin after the train passage in milliseconds
+    ///
+    /// Enforces that the path used by the train should be free and
+    /// available at least that many milliseconds after its passage.
+    #[serde(default)]
+    pub time_gap_after: u64,
+    /// Can be a percentage `X%`, a time in minutes per 100 kilometer `Xmin/100km`
+    #[serde(default)]
+    #[schema(value_type = Option<String>, example = json!(["5%", "2min/100km"]))]
+    pub margin: Option<MarginValue>,
+    /// Total mass of the consist in kg
+    pub total_mass: Option<f64>,
+    /// Total length of the consist in meters
+    pub total_length: Option<f64>,
+    /// Maximum speed of the consist in km/h
+    pub max_speed: Option<f64>,
+    pub loading_gauge_type: Option<LoadingGaugeType>,
+}
+
+impl Request {
+    /// Returns the earliest time that has been set on any step
+    pub(super) fn get_earliest_step_time(&self) -> DateTime<Utc> {
+        // Get the earliest time that has been specified for any step
+        self.start_time
+            .or_else(|| {
+                self.steps
+                    .iter()
+                    .flat_map(|step| step.timing_data.iter())
+                    .map(|data| {
+                        data.arrival_time
+                            - Duration::milliseconds(data.arrival_time_tolerance_before as i64)
+                    })
+                    .next()
+            })
+            .expect("No time specified for stdcm request")
+    }
+
+    /// Returns the earliest tolerance window that has been set on any step
+    fn get_earliest_step_tolerance_window(&self) -> u64 {
+        // Get the earliest time window that has been specified for any step, if maximum_run_time is not none
+        self.steps
+            .iter()
+            .flat_map(|step| step.timing_data.iter())
+            .map(|data| data.arrival_time_tolerance_before + data.arrival_time_tolerance_after)
+            .next()
+            .unwrap_or(0)
+    }
+
+    /// Returns the request's total stop time
+    fn get_total_stop_time(&self) -> u64 {
+        self.steps
+            .iter()
+            .map(|step: &PathfindingItem| step.duration.unwrap_or_default())
+            .sum()
+    }
+
+    // Returns the maximum departure delay for the train.
+    pub(super) fn get_maximum_departure_delay(&self, simulation_run_time: u64) -> u64 {
+        self.maximum_departure_delay
+            .unwrap_or(simulation_run_time + self.get_earliest_step_tolerance_window())
```

Comment:
That feels wrong, I'd expect this instead:
```suggestion
            .unwrap_or(self.get_earliest_step_tolerance_window())
```

---

@Erashin this function has just been moved here and you wrote the previous version, do you remember if there's a reason to add the simulation run time there?

Reply:
In the case where the departure is provided, that is indeed the case. However, if it's the end that is provided, it will always fail if the simulation run time is a lot higher than the tolerance window.. the best would be to differentiate both cases :
- if the departure is provided, max departure delay should be the tolerance window
- if it's the end, it should be as specified right now.